async_timeout/__init__.py

```diff
@@ -179,30 +179,31 @@ def shift_to(self, deadline: float) -> None:
             else:
                 # state is ENTER
                 raise asyncio.CancelledError
-        self._timeout_handler = self._loop.call_at(
-            deadline, self._on_timeout, self._task
-        )
+        self._timeout_handler = self._loop.call_at(deadline, self._on_timeout)
 
     def _do_enter(self) -> None:
         if self._state != _State.INIT:
-            raise RuntimeError("invalid state {}".format(self._state.value))
+            raise RuntimeError(f"invalid state {self._state.value}")
         self._state = _State.ENTER
 
     def _do_exit(self, exc_type: Type[BaseException]) -> None:
+        self._task = None  # Early drop circular reference
         if exc_type is asyncio.CancelledError and self._state == _State.TIMEOUT:
             self._timeout_handler = None
             raise asyncio.TimeoutError
-        # timeout is not expired
+        # timeout is not expired, propagate the exception
         self._state = _State.EXIT
         self._reject()
         return None
 
-    def _on_timeout(self, task: "asyncio.Task[None]") -> None:
+    def _on_timeout(self) -> None:
         # See Issue #229 and PR #230 for details
-        if task._fut_waiter and task._fut_waiter.cancelled():  # type: ignore[attr-defined]  # noqa: E501
-            return
-        task.cancel()
-        self._state = _State.TIMEOUT
+        self._loop.call_soon(self._cancel)
```

Comment:
I find this much harder to reason about, I think it's a more risky than the other solutions.

The reason this somewhat works is that it changes the call order in the before case to:
-> t.cancel()
-> Timeout._on_timeout()
-> Timeout._on_exit()
-> Timeout._cancel()  # This is a noop, as it's already exited

I'm not really sure that provides any guarantees, as it's just delaying the cancellation, which happens to be after the exit has already happened.

In the after case, this is not fixed at all (just like all the other attempts), in this case the execution looks like:
-> Timeout._on_timeout()
-> t.cancel()
-> Timeout._cancel()  # Again, we've triggered 2 cancellations in a row
-> Timeout._do_exit()  # No way to know there were 2 cancellations.